server/src/main/java/org/elasticsearch/action/bulk/BulkOperation.java

```diff
@@ -239,33 +381,135 @@ public void onFailure(Exception e) {
+                    if (failureStoreReference != null) {
+                        addDocumentToRedirectRequests(request, e, failureStoreReference);
+                    }
+                    addFailure(docWriteRequest, request.id(), indexName, e);
                 }
+                completeShardOperation();
+            }
+
+            private void completeShardOperation() {
+                // Clear our handle on the cluster state to allow it to be cleaned up
+                clusterState = null;
                 releaseOnFinish.close();
             }
         });
     }
 
-    private boolean handleBlockExceptions(ClusterState state) {
+    /**
+     * Determines if the write request can be redirected if it fails. Write requests can be redirected IFF they are targeting a data stream
+     * with a failure store and are not already redirected themselves. If the document can be redirected, the data stream name to use for
+     * the redirection is returned.
+     *
+     * @param docWriteRequest the write request to check
+     * @param metadata cluster state metadata for resolving index abstractions
+     * @return a data stream name if the write request points to a data stream that has the failure store enabled,
+     *     or {@code null} if it does
+     */
+    private static String getRedirectTarget(DocWriteRequest<?> docWriteRequest, Metadata metadata) {
+        // Feature flag guard
+        if (DataStream.isFailureStoreEnabled() == false) {
+            return null;
+        }
+        // Do not resolve a failure store for documents that were already headed to one
+        if (docWriteRequest instanceof IndexRequest indexRequest && indexRequest.isWriteToFailureStore()) {
+            return null;
+        }
+        // If there is no index abstraction, then the request is using a pattern of some sort, which data streams do not support
+        IndexAbstraction ia = metadata.getIndicesLookup().get(docWriteRequest.index());
+        if (ia == null) {
+            return null;
+        }
+        if (ia.isDataStreamRelated()) {
+            // The index abstraction could be an alias. Alias abstractions (even for data streams) only keep track of which _index_ they
+            // will write to, not which _data stream_.
+            // We work backward to find the data stream from the concrete write index to cover this case.
+            Index concreteIndex = ia.getWriteIndex();
+            IndexAbstraction writeIndexAbstraction = metadata.getIndicesLookup().get(concreteIndex.getName());
+            DataStream parentDataStream = writeIndexAbstraction.getParentDataStream();
+            if (parentDataStream != null && parentDataStream.isFailureStore()) {
+                // Keep the data stream name around to resolve the redirect to failure store if the shard level request fails.
+                return parentDataStream.getName();
+            }
```

Comment:
Food for thought, I think we might need this more often as a follow up PR. Maybe should add this in the `IndexAbstraction`, I have a feeling it's not the first time we do something like that.